libraries/chain/controller.cpp

```diff
@@ -3994,48 +4009,45 @@ std::optional<signed_block_header> controller::fetch_block_header_by_id( const b
 }
 
 signed_block_ptr controller::fetch_block_by_number( uint32_t block_num )const  { try {
-   auto blk_state = fetch_block_state_by_number( block_num );
-   if( blk_state ) {
-      return blk_state->block;
-   }
+   auto fetch_block = [&](auto& fork_db, auto& head) -> signed_block_ptr {
+      auto bsp = fork_db.search_on_branch( fork_db.head()->id(), block_num);
+      if (bsp) return bsp->block;
+      return {};
+   };
+   auto b = my->block_data.apply<signed_block_ptr>(fetch_block);
+   if (b)
+      return b;
 
    return my->blog.read_block_by_num(block_num);
 } FC_CAPTURE_AND_RETHROW( (block_num) ) }
 
 std::optional<signed_block_header> controller::fetch_block_header_by_number( uint32_t block_num )const  { try {
-   auto blk_state = fetch_block_state_by_number( block_num );
-   if( blk_state ) {
-      return blk_state->header;
-   }
+   auto fetch_block = [&](auto& fork_db, auto& head) -> signed_block_ptr {
+      auto bsp = fork_db.search_on_branch( fork_db.head()->id(), block_num);
+      if (bsp) return bsp->block;
+      return {};
+   };
+   auto b = my->block_data.apply<signed_block_ptr>(fetch_block);
```

Comment:
Because this is duplicated here and above in `fetch_block_by_number`, and might be useful elsewhere, I think we should define a function in `struct block_data_t `:

```
   signed_block_ptr fetch_block(uint32_t block_num) const {
      return std::visit([&](const auto& bd) -> signed_block_ptr {
         auto bsp = bd.fork_db.search_on_branch(bd.fork_db.head()->id(), block_num);
         if (bsp) return bsp->block;
         return {};
      }, v);
   }
```

Reply:
Done.